flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartition.java

```diff
@@ -152,7 +152,10 @@ private void writeAndCloseBufferConsumer(BufferConsumer bufferConsumer) throws I
 				if (canBeCompressed(buffer)) {
 					final Buffer compressedBuffer = parent.bufferCompressor.compressToIntermediateBuffer(buffer);
 					data.writeBuffer(compressedBuffer);
-					compressedBuffer.recycleBuffer();
+					// release the intermediate buffer
+					if (compressedBuffer != buffer) {
+						compressedBuffer.recycleBuffer();
```

Comment:
Actually this should be a separate bug out of this ticket scope, so it is better to submit this fix separately.

In addition this issue let me think why it was not exposed by tests before. I think the following `bufferConsumer.close()` somehow allows the inconsistent logics to bring potential risks.  In implementing of `BufferConsumer#close()` it would firstly check whether the backed buffer was released or not. If not released yet, then it would call `Buffer#recycleBuffer`, otherwise it does nothing. 

I am not sure why we define such loose semantics for `BufferConsumer#close`, rather than make it as a strict semantic. Do you have any inputs here @pnowojski ?

Reply:
+1 this must be an independent bug fix/jira ticket, with it's own test coverage.

Wasn't it discovered because this is a rare case, that compression couldn't reduce the size of the data?

>  I think the following bufferConsumer.close() somehow allows the inconsistent logics to bring potential risks. 

What do you mean?

> I am not sure why we define such loose semantics for BufferConsumer#close, rather than make it as a strict semantic.

The if check might have been added to implement `Closeable` interface. Note the difference between `Closeable` and `AutoCloseable`:
```
public interface Closeable extends AutoCloseable {

    /**
     * Closes this stream and releases any system resources associated
     * with it. If the stream is already closed then invoking this
     * method has no effect.
```
`Closeable` is an easier interface to use, as user can just do not care if it was closed before or not. I'm not sure if we strictly need this though. Maybe we do or maybe I've implemented it that way because I prefer `Closeable` more ;) ).

Reply:
It is mainly because before calling the `BufferConsumer#close`, the results are both correct when `recycleBuffer()` was called once or twice. 

- If `recycleBuffer()` was called once, the `close()` would only trigger recycle buffer internally to make it released.
- If `recycleBuffer()` was called twice, the `close()` would do nothing because the buffer was already released.

Before changes, the recycle was called only once,  and after changes it was called twice. So the potential inconsistent logics are hidden. And that is my concern of not very strict semantic of `close` method which would allow the user making mistake in the middle of the process.

Reply:
Thanks for the review and comment. I would like to create a new ticket for this.